Documentation for NodeLease feature being promoted to GA (#17189)
kubernetes/website: content/en/docs/concepts/architecture/nodes.md

ConditionUnknown and 5m after that to start evicting pods.) The node controller
checks the state of each node every `--node-monitor-period` seconds.

#### Heartbeats

> Each Node  has an associated `Lease` object in `kube-node-lease` namespace.

Reviewer: ```suggestion
Each Node has an associated Lease object in the `kube-node-lease` {{< glossary_tooltip term_id="namespace" text="namespace">}}.
```

Author: done

> It is periodically renewed by the kubelet and both NodeStatus and the Lease

Reviewer: ```suggestion
The kubelet updates its Lease object frequently to show that its Node is healthy. In the control plane, both NodeStatus and the Lease
```

Author: done

are treated as heartbeats from the node.

> Node leases are renewed frequently while NodeStatus is reported from node to

Reviewer: If accepting the previous suggestion, how about this wording?
```markdown
The kubelet updates NodeStatus either when there is change in status, or
if there has been no update for a configured interval. The default interval
for NodeStatus updates is 5 minutes (much longer than the 40 second
default timeout for unreachable nodes).
```

Author: done

master only where there is some change or enough time has passed (default is
5 minutes, which is longer than the default timeout of 40 seconds for

> unreachable nodes). Compared to the Node resource, the Lease object is

Reviewer: > Compared to the Node resource, the Lease object is lightweight. The Lease resource improves the performance of the node heartbeats as the cluster scales.

As NodeLease is GA (once this is merged, at least), and the oldest supported Kubernetes version will have NodeLease enabled by default, it's less important to explain the older method. I think it's OK to assume that readers are using NodeLease and that it pretty much “just works” for them.
I would cut out these 2 sentences.

Author: I'm not sure - this shows why this is even needed.

ConditionUnknown and 5m after that to start evicting pods.) The node controller
checks the state of each node every `--node-monitor-period` seconds.

#### Heartbeats

> 

Reviewer 2: Hi. I looked at this again. Added a few sentences at the beginning and moved a few more around.
Also created a small list out of the kubelet update info. Not sure about the Lease update interval?
Feel free to use what you like.

Heartbeats, between a node and the control plane, help determine the state of a node.
A heartbeat represents the Node resource `NodeStatus` and the (Lease object) [docs/reference/generated/kubernetes-api/v1.16/#lease-v1-coordination-k8s-io]. Each Node has an associated `Lease` object in the `kube-node-lease` {{< glossary_tooltip term_id="namespace" text="namespace">}}. Compared to the `Node` resource, the `Lease` object is lightweight.

The `kubelet` is responsible for creating and updating the `NodeStatus` and `Lease`.

- The `kubelet` updates the `NodeStatus` either when there is change in status, or
if there has been no update for a configured interval. The default interval
for `NodeStatus` updates is 5 minutes (much longer than the 40 second default timeout for unreachable nodes).
- The `kubelet` creates and updates the `Lease` object. The frequency of the `Lease` updates is independent from the frequency of the `NodeStatus` updates.

With a small footprint heartbeat, the load of maintaining the state of a node is decreased. The `Lease` resource improves the performance of the node heartbeats as the cluster scales.

Author: Thanks - I really like it.
I just modified it slightly - please take a look if that still makes sense to you and suggest improvements.

ConditionUnknown and 5m after that to start evicting pods.) The node controller
checks the state of each node every `--node-monitor-period` seconds.

#### Heartbeats

Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.
There are two forms of heartbeats: updates of `NodeStatus` and the
(Lease object)[docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io].

> Each Node has an associated `Lease` object in the `kube-node-lease`

Reviewer 2: Changes seem fine. Looks like I introduced a formatting error. Pls correct line 180:
[Lease object](/docs/reference/generated/kubernetes-api/{{< latest-version >}}/#lease-v1-coordination-k8s-io).

Author: Done - PTAL

Reviewer 2: /lgtm

> Each Node has an associated `Lease` object in the `kube-node-lease`

Reviewer 2: These changes look good. If you are cleaning up nits ...
- There is one more Lease to clean up the backticks on line 181.
- Also, I _think_ command tools (kubelet, kubectl) typically are quoted with backticks, throughout the docs.
  For example: https://kubernetes.io/docs/tasks/administer-cluster/out-of-resource/

Reviewer: `kubectl` is 100% a command line tool. `kubelet`, whilst also an executable,  isn't a tool that people _usually_ run in a command line, at least not during typical cluster operation. I see `kubelet` fitting into the same slot as `kube-scheduler`, `kube-controller-manager`, and `kube-proxy`.

(I've used backticks for those names in this very comment but in the docs I typically wouldn't).

Author: Fixed the Lease.

I agree with @sftim that kubelet is in the same category as kube-scheduler, etc.
[in this file kubelet is not in backsticks anywhere]

Reviewer 2: Hello @wojtek-t . It feels like these changes should be ready 😀 . Okay to override these comments, but if you have the energy for one more change, I'd suggest splitting/adjusting lines 183 - 184.

Lease is a lightweight resource, which improves the performance of the node heartbeats as the cluster scales.
OR
Lease is a smaller resource than the Node, which improves the performance of the node heartbeats as the cluster scales.

Line 186: The kubelet is responsible for creating and updating the `NodeStatus` and a Lease object.

Author: Done.

{{< glossary_tooltip term_id="namespace" text="namespace">}}.
Compared to the Node resource, the Lease is a lightweight resource, which improves
the performance of the node heartbeats as the cluster scales.

> The `kubelet` is responsible for creating and updating `NodeStatus` and `Lease`.

Reviewer: Late, nit tweaks:
```suggestion
The kubelet is responsible for creating and updating `NodeStatus` and a Lease.
```

Author: done

> - The `kubelet` updates the `NodeStatus` either when there is change in status,

Reviewer: ```suggestion
- The kubelet updates the `NodeStatus` either when there is change in status,
```

Author: done

  or if there has been no update for a configured interval. The default interval
  for `NodeStatus` updates is 5 minutes (much longer than the 40 second default
  timeout for unreachable nodes).

> - The `kubelet` creates and then updates the `Lease` object every 10 seconds

Reviewer: ```suggestion
- The kubelet creates and then updates its Lease object every 10 seconds
```

Author: done

>   (the default update interval). `Lease` updates occur independently from the

Reviewer: ```suggestion
(the default update interval). Lease updates occur independently from the
```

Author: done

ConditionUnknown and 5m after that to start evicting pods.) The node controller
checks the state of each node every `--node-monitor-period` seconds.

#### Heartbeats

Heatbeats, sent by Kubernetes nodes, help determining their state.

> There are two forms of hearbeats: updates of `NodeStatus` an

Reviewer 2: @wojtek-t , Some more suggestions+nits.
line 171- 174: Somewhat related: these lines seem to be a repeat of lines 187 -190?
line 178: Spelling, wording: Heartbeats, sent by Kubernetes nodes [to the control plane], help determine the health [or availability] of a node.
line 179: Spelling, wording: There are two forms of heartbeats: updates containing `NodeStatus` and updates containing the (Lease object){ ...].
line 181: in the `kube-node-lease` ...
line 191: Wording to make concise: The `kubelet` updates the `Lease` object every 10 seconds (the default update interval).
line 195 - 197: Possibly move content, reduce, and merge with line 183 (@sftim , comments on including this statement?):
`Lease` is a lightweight resource, which improves the performance
of the node heartbeats as the cluster scales.

Author: > line 171- 174: Somewhat related: these lines seem to be a repeat of lines 187 -190?

I disagree - 171-174 is talking what nodecontroller is doing based on signals it get
lines 187-190 is discussing what those signals are

> line 179: Spelling, wording: There are two forms of heartbeats: updates containing NodeStatus and updates containing the (Lease object){ ...].

Those aren't updates that contain something - in the REST api these are updates of objects.

Other comments should addressed.

Reviewer 2: @wojtek-t , OK. Almost there.

line 178: pls. take out square brackets -- I intended you to choose the words that fit best (where are the heartbeats sent? to the control plane, _availability_ vs. _health_):
Heartbeats, sent by Kubernetes nodes, help determine the availability of a node.
line 180: s/hearbeats/heartbeats/
line 180: If this makes sense, replace _an- with _and the_: updates of `NodeStatus` and the ...
line 193: Reverting back to two sentences, as it seems easier to read
The `kubelet` creates and then updates the `Lease` object every 10 seconds (the default update interval).
`Lease` updates occur independently from the `NodeStatus` updates.

Author: Thanks a lot for patience. Done.

ConditionUnknown and 5m after that to start evicting pods.) The node controller
checks the state of each node every `--node-monitor-period` seconds.

> Each node has an associated `Lease` object in `kube-node-lease` namespace.

Reviewer: nit
```suggestion
Each Node has an associated `Lease` object in the `kube-node-lease` namespace.
```

Reviewer: Alternatively (see other comment) omit this line.

It is periodically renewed by Kubelet and both NodeStatus and node lease
are treated as heartbeats from the node.
Node leases are renewed frequently while NodeStatus is reported from node to
master only where there is some change or enough time has passed (default is
5 minutes, which is longer than the default timeout of 40 seconds for

> unreachable nodes). Since Lease is much more lighweigh object than Node,

Reviewer: 1. Typo: “lighweigh”
2. This page is about Nodes, and the reader could be totally new to Kubernetes. This page is a lot of detail to explain what a Node is. If anything, I'd prefer to slim it further.

Reviewer: How about a heading: “Node leases” plus 100 words on the control plane issuing Node leases to keep track of Node health / failures.

Does NodeStatus even need mentioning here?

Author: I think it does - the reason is that for the last 5 years, NodeStatus was the only signal for heartbeat. Now this is only treated as an additional one.

Author: That said, I tried to rephrase it a bit (applied suggestions by @kbhawkey ) and slightly better organize.
